crates/common/src/ether/http_or_ws_or_ipc.rs

```diff
@@ -0,0 +1,165 @@
+// This will connect to [`Http`] if the rpc_url contains 'http', to [`Ws`] if it contains 'ws',
+// otherwise it'll default to [`Ipc`].
+impl HttpOrWsOrIpc {
+    pub async fn connect(rpc_url: &str) -> Result<Self, HttpOrWsOrIpcError> {
+        let this = if rpc_url.to_lowercase().contains("http") {
+            Self::Http(Http::from_str(rpc_url).unwrap())
+        } else if rpc_url.to_lowercase().contains("ws") {
+            Self::Ws(Ws::connect(rpc_url).await?)
+        } else {
+            Self::Ipc(Ipc::connect(rpc_url).await?)
+        };
+        Ok(this)
+    }
+}
+
+// Next, the most important step: implement [`JsonRpcClient`].
+//
+// For this implementation, we simply delegate to the wrapped transport and return the
+// result.
+//
+// Note that we are using [`async-trait`](https://docs.rs/async-trait) for asynchronous
+// functions in traits, as this is not yet supported in stable Rust; see:
+// <https://blog.rust-lang.org/inside-rust/2022/11/17/async-fn-in-trait-nightly.html>
+#[async_trait]
+impl JsonRpcClient for HttpOrWsOrIpc {
+    type Error = HttpOrWsOrIpcError;
+
+    async fn request<T, R>(&self, method: &str, params: T) -> Result<R, Self::Error>
+    where
+        T: Debug + Serialize + Send + Sync,
+        R: DeserializeOwned + Send, {
+        let res = match self {
+            Self::Ws(ws) => JsonRpcClient::request(ws, method, params).await?,
+            Self::Ipc(ipc) => JsonRpcClient::request(ipc, method, params).await?,
+            Self::Http(http) => JsonRpcClient::request(http, method, params).await?,
+        };
+        Ok(res)
+    }
+}
+
+// We can also implement [`PubsubClient`], since both `Ws` and `Ipc` implement it, by
+// doing the same as in the `JsonRpcClient` implementation above.
+// Trying to subscribe on a `Http` will panic.
+impl PubsubClient for HttpOrWsOrIpc {
+    // Since both `Ws` and `Ipc`'s `NotificationStream` associated type is the same,
+    // we can simply return one of them.
+    // In case they differed, we would have to create a `HttpOrWsOrIpcNotificationStream`,
+    // similar to the error type.
+    type NotificationStream = <Ws as PubsubClient>::NotificationStream;
+
+    fn subscribe<T: Into<U256>>(&self, id: T) -> Result<Self::NotificationStream, Self::Error> {
+        let stream = match self {
+            Self::Ws(ws) => PubsubClient::subscribe(ws, id)?,
+            Self::Ipc(ipc) => PubsubClient::subscribe(ipc, id)?,
+            HttpOrWsOrIpc::Http(_) => unreachable!("Http RPC cannot be used for subscriptions!"),
+        };
+        Ok(stream)
+    }
+
+    fn unsubscribe<T: Into<U256>>(&self, id: T) -> Result<(), Self::Error> {
+        match self {
+            Self::Ws(ws) => PubsubClient::unsubscribe(ws, id)?,
+            Self::Ipc(ipc) => PubsubClient::unsubscribe(ipc, id)?,
+            HttpOrWsOrIpc::Http(_) => unreachable!("Http RPC cannot be used for subscriptions!"),
+        };
+        Ok(())
+    }
+}
+
+#[tokio::test]
+async fn test_subscription() {
```

Comment:
lets put all tests in a

```
#[cfg(test)]
mod tests { .. }
```

also, do we know why this test fails?

Reply:
It was failing, because to use anvil you need to install it on the machine additionally. I changed to use an external ws link, hopefully it's reliable enough. Ethers.rs has an additional step in github actions to install anvil, but I imagine we don't want to add this overhead until we use it more widely.